fixes inconsistent lighting ci failure in icebox

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
![Screenshot_2045](https://user-images.githubusercontent.com/15794172/171071869-9f74bdb5-be35-4dae-8ca5-5a0e3586c5b6.png)
happened across this error when testing icebox, remembered something about this being a ci failure. turns out it was because `pixel_turf` was null, which happens get_turf_pixel() returns null here in light_source/proc/update_corners():
```
if (isturf(top_atom))
		if (source_turf != top_atom)
			source_turf = top_atom
			pixel_turf = source_turf
			update = TRUE
	else if (top_atom.loc != source_turf)
		source_turf = top_atom.loc
		pixel_turf = get_turf_pixel(top_atom)
		update = TRUE
	else
		var/pixel_loc = get_turf_pixel(top_atom)
		if (pixel_loc != pixel_turf)
			pixel_turf = pixel_loc
			update = TRUE
```

when does get_turf_pixel() return null? turns out, top_atom in this case was a colossus mob which is taller than a tile, and it happened to spawn at y = 255. so get_turf_pixel(colossus at the top of the map) was trying to find the turf at locate(colossus.x, colossus.y + top of his tall ass sprite, colossus.z), which is > 255, so it returned null and made lighting objects runtime. now it clamps the values to the bounds of the map and doesnt runtime in the same way

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
fixes ci
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
fixes spurious lighting cli failure in icebox

Kylerace committed:
adds a unit test

Kylerace commented on code/__HELPERS/turfs.dm:
```diff
@@ -229,8 +229,8 @@ Turf and target are separate in case you want to teleport some distance from a t
 		return null
-	var/final_x = atom_turf.x + rough_x
-	var/final_y = atom_turf.y + rough_y
+	var/final_x = clamp(atom_turf.x + rough_x, 1, world.maxx)
+	var/final_y = clamp(atom_turf.y + rough_y, 1, world.maxy)
```
```suggestion
	var/final_x = atom_turf.x + rough_x
	var/final_y = atom_turf.y + rough_y
```
testing to see if the added unit test fails under current behavior

Kylerace replied:
it doesnt, but it should

Kylerace replied:
ok now that the test actually RUNS it does fail

Kylerace committed:
Revert "Update code/__HELPERS/turfs.dm"

This reverts commit e2f016241ff62a931b9aa5cd544c601fffd43056.

Kylerace committed:
Revert "Revert "Update code/__HELPERS/turfs.dm""

This reverts commit 149d90dd41149607b3a2a84efce3cb9c4b5d20ce.

Kylerace committed:
Update code/__HELPERS/turfs.dm

optimumtact commented:
we might benefit from a unit test which just spawns one of every atom on the corner of maps, we could parallelize it by using multiple z's and spawning 4 to a z

optimumtact commented:
also how did you replicate this?

optimumtact commented:
also this isn't a spurious failure, it's a valid one???

>adj. not being what it purports to be; false or fake.

Mothblocks commented:
@optimumtact i thought spurious meant random, i'll rename the project

Kylerace commented:
Spurious project name

Kylerace commented:
> also how did you replicate this?

Spawning a colossus at y=255 on icebox replicates it every time

Kylerace commented:
> we might benefit from a unit test which just spawns one of every atom on the corner of maps, we could parallelize it by using multiple z's and spawning 4 to a z

Only the bounds of the atom should matter

optimumtact commented:
> > we might benefit from a unit test which just spawns one of every atom on the corner of maps, we could parallelize it by using multiple z's and spawning 4 to a z
> 
> Only the bounds of the atom should matter

Testing is a good way to find out



> > also how did you replicate this?
> 
> Spawning a colossus at y=255 on icebox replicates it every time

sorry bad use of terminology on my prat, I meant more how did you find the issue originally.

Kylerace committed:
i cant actually test for the east and south directions

Kylerace committed:
fuck it ill just have him overhand the north side

Kylerace commented:
>i cant even get my test to fail

Mothblocks commented:
the importance of test driven development :)

Kylerace commented:
the tests are driving me mad

jlsnow301 commented:
Fixes #63831

Kylerace committed:
maybe this gets it to run

Kylerace committed:
ok THIS will get it to run

Kylerace committed:
no, THIS will get it to run

Kylerace committed:
now we know the test is failed every time without the fix. readds fix

Mothblocks commented on code/modules/unit_tests/get_turf_pixel.dm:
```diff
@@ -0,0 +1,15 @@
+	var/mob/living/simple_animal/hostile/megafauna/colossus/long_larry = allocate(/mob/living/simple_animal/hostile/megafauna/colossus, north)
+	TEST_ASSERT(get_turf_pixel(long_larry) != null, "get_turf_pixel() isnt clamping a mob whos sprite is above the bounds of the world inside of the map.")
+
+
+
+
+
```
bro remove the oasis

Kylerace replied:
removed oasis in commit [`b2b5c04` (#67430)](https://github.com/tgstation/tgstation/pull/67430/commits/b2b5c044880bb4aa797a8f5f9f1dc718b5328708)

Mothblocks commented on code/modules/unit_tests/get_turf_pixel.dm:
```diff
@@ -0,0 +1,15 @@
+	//we need long larry to peek over the top edge of the earth
+	var/turf/north = locate(1, world.maxy, run_loc_floor_bottom_left.z)
+
+	///hes really long, so hes really good at peaking over the edge of the map
```
```suggestion
	//hes really long, so hes really good at peaking over the edge of the map
```

Not a doc comment

Kylerace committed:
today is gonna be the day that they're gonna throw it back to you.

And by now, you should've somehow realised what you gotta do
I don't believe that anybody feels the way I do about you now
And backbeat, the word is on the street that the fire in your heart is out
I'm sure you've heard it all before, but you never really had a doubt
I don't believe that anybody feels the way I do about you now
And all the roads we have to walk are winding
And all the lights that lead us there are blinding
There are many things that I would like to say to you, but I don't know how
Because maybe
You're gonna be the one that saves me
And after all
You're my wonderwall
Today was gonna be the day, but they'll never throw it back to you
And by now, you should've somehow realised what you're not to do
I don't believe that anybody feels the way I do about you now
And all the roads that lead you there were winding
And all the lights that light the way are blinding
There are many things that I would like to say to you, but I don't know how
I said maybe
You're gonna be the one that saves me
And after all
You're my wonderwall
I said maybe (I said maybe)
You're gonna be the one that saves me
And after all
You're my wonderwall
I said maybe (I said maybe)
You're gonna be the one that saves me (saves me)
You're gonna be the one that saves me (saves me)
You're gonna be the one that saves me (saves me)

Kylerace committed:
Update code/modules/unit_tests/get_turf_pixel.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Kylerace commented on code/modules/unit_tests/get_turf_pixel.dm:
```diff
@@ -0,0 +1,11 @@
+
+	//hes really long, so hes really good at peaking over the edge of the map
+	var/mob/living/simple_animal/hostile/megafauna/colossus/long_larry = allocate(/mob/living/simple_animal/hostile/megafauna/colossus, north)
+	TEST_ASSERT(get_turf_pixel(long_larry) != null, "get_turf_pixel() isnt clamping a mob whos sprite is above the bounds of the world inside of the map.")
```
```suggestion
	TEST_ASSERT(istype(get_turf_pixel(long_larry), /turf), "get_turf_pixel() isnt clamping a mob whos sprite is above the bounds of the world inside of the map.")
```
just realized that checking for null is bad since someone can later change it to return FALSE. we just need to test whether it returns a turf which it should

Kylerace committed:
Update code/modules/unit_tests/get_turf_pixel.dm
